// file-box/dist/src/file-box.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 *
 * File Box
 * https://github.com/huan/file-box
 *
 * 2018 Huan LI <zixia@zixia.net>
 *
 */
const fs = require("fs");
const nodePath = require("path");
const mime = require("mime");
const stream_1 = require("stream");
const config_1 = require("./config");
const file_box_type_1 = require("./file-box.type");
const misc_1 = require("./misc");
const EMPTY_META_DATA = Object.freeze({});
class FileBox {
    /**
     *
     * Static Properties
     *
     */
    /**
     * Alias for `FileBox.fromUrl()`
     *
     * @alias fromUrl()
     */
    static fromUrl(url, name, headers) {
        if (!name) {
            const parsedUrl = nodePath.parse(url);
            name = parsedUrl.base;
        }
        const options = {
            type: file_box_type_1.FileBoxType.Url,
            name,
            url,
            headers,
        };
        return new FileBox(options);
    }
    /**
     * Alias for `FileBox.fromFile()`
     *
     * @alias fromFile
     */
    static fromFile(path, name) {
        if (!name) {
            name = nodePath.parse(path).base;
        }
        const options = {
            type: file_box_type_1.FileBoxType.File,
            name,
            path: path,
        };
        return new FileBox(options);
    }
    static fromStream(stream, name) {
        const options = {
            type: file_box_type_1.FileBoxType.Stream,
            stream,
            name,
        };
        return new FileBox(options);
    }
    static fromBuffer(buffer, name) {
        const options = {
            type: file_box_type_1.FileBoxType.Buffer,
            name,
            buffer,
        };
        return new FileBox(options);
    }
    static fromBase64(base64, name) {
        const options = {
            type: file_box_type_1.FileBoxType.Buffer,
            name,
            buffer: Buffer.from(base64, 'base64'),
        };
        return new FileBox(options);
    }
    /**
     * dataURL: `data:image/png;base64,${base64Text}`,
     */
    static fromDataURL(dataUrl, name) {
        return FileBox.fromBase64(misc_1.dataUrlToBase64(dataUrl), name);
    }
    static version() {
        return config_1.VERSION;
    }
    get metadata() {
        if (this._metadata) {
            return this._metadata;
        }
        return EMPTY_META_DATA;
    }
    set metadata(data) {
        if (this._metadata) {
            throw new Error('metadata can not be modified after set');
        }
        this._metadata = Object.assign({}, data);
        Object.freeze(this._metadata);
    }
    constructor(fileOrOptions) {
        let options;
        if (typeof fileOrOptions === 'string') {
            /**
             * Default to Local File
             */
            options = {
                type: file_box_type_1.FileBoxType.File,
                name: fileOrOptions,
                path: nodePath.resolve(fileOrOptions),
            };
        }
        else {
            options = fileOrOptions;
        }
        // Only keep `basename` in this.name
        this.name = nodePath.basename(options.name);
        this.boxType = options.type;
        this.mimeType = mime.getType(this.name) || undefined;
        switch (options.type) {
            case file_box_type_1.FileBoxType.Buffer:
                if (!options.buffer) {
                    throw new Error('no buffer');
                }
                this.buffer = options.buffer;
                break;
            case file_box_type_1.FileBoxType.File:
                if (!options.path) {
                    throw new Error('no path');
                }
                this.localPath = options.path;
                break;
            case file_box_type_1.FileBoxType.Url:
                if (!options.url) {
                    throw new Error('no url');
                }
                this.remoteUrl = options.url;
                if (options.headers) {
                    this.headers = options.headers;
                }
                break;
            case file_box_type_1.FileBoxType.Stream:
                if (!options.stream) {
                    throw new Error('no stream');
                }
                this.stream = options.stream;
                break;
            default:
                throw new Error('unknown type');
        }
    }
    version() {
        return config_1.VERSION;
    }
    toString() {
        return [
            'FileBox#',
            file_box_type_1.FileBoxType[this.boxType],
            '<',
            this.name,
            '>',
        ].join('');
    }
    toJSON() {
        throw new Error('WIP');
    }
    ready() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.boxType === file_box_type_1.FileBoxType.Url) {
                yield this.syncRemoteName();
            }
        });
    }
    /**
     * @todo use http.get/gets instead of Request
     */
    syncRemoteName() {
        return __awaiter(this, void 0, void 0, function* () {
            /**
             * https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition
             *  > Content-Disposition: attachment; filename="cool.html"
             */
            if (this.boxType !== file_box_type_1.FileBoxType.Url) {
                throw new Error('type is not Remote');
            }
            if (!this.remoteUrl) {
                throw new Error('no url');
            }
            const headers = yield misc_1.httpHeadHeader(this.remoteUrl);
            const filename = misc_1.httpHeaderToFileName(headers);
            if (filename) {
                this.name = filename;
            }
            if (!this.name) {
                throw new Error('no name');
            }
            this.mimeType = headers['content-type'] || mime.getType(this.name) || undefined;
        });
    }
    pipe(destination) {
        this.toStream().then(stream => stream.pipe(destination));
        return destination;
    }
    toStream() {
        return __awaiter(this, void 0, void 0, function* () {
            let stream;
            switch (this.boxType) {
                case file_box_type_1.FileBoxType.Buffer:
                    stream = this.transformBufferToStream();
                    break;
                case file_box_type_1.FileBoxType.File:
                    stream = this.transformFileToStream();
                    break;
                case file_box_type_1.FileBoxType.Url:
                    stream = yield this.transformUrlToStream();
                    break;
                case file_box_type_1.FileBoxType.Stream:
                    if (!this.stream) {
                        throw new Error('no stream');
                    }
                    stream = this.stream;
                    break;
                default:
                    throw new Error('not supported FileBoxType: ' + file_box_type_1.FileBoxType[this.boxType]);
            }
            return stream;
        });
    }
    /**
     * https://stackoverflow.com/a/16044400/1123955
     */
    transformBufferToStream() {
        const bufferStream = new stream_1.PassThrough();
        bufferStream.end(this.buffer);
        return bufferStream;
    }
    transformFileToStream() {
        const filePath = this.localPath;
        if (!filePath) {
            throw new Error('no url(path)');
        }
        return fs.createReadStream(filePath);
    }
    transformUrlToStream() {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                if (this.remoteUrl) {
                    misc_1.httpStream(this.remoteUrl, this.headers)
                        .then(resolve);
                }
                else {
                    reject(new Error('no url'));
                }
            });
        });
    }
    /**
     * save file
     *
     * @param filePath save file
     */
    toFile(filePath, overwrite = false) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.boxType === file_box_type_1.FileBoxType.Url) {
                if (!this.mimeType || !this.name) {
                    yield this.syncRemoteName();
                }
            }
            const fullFilePath = nodePath.resolve(filePath || this.name);
            const exist = yield new Promise(resolve => fs.exists(fullFilePath, resolve));
            if (exist && !overwrite) {
                throw new Error(`FileBox.toFile(${fullFilePath}): file exist. use FileBox.toFile(${fullFilePath}, true) to force overwrite.`);
            }
            const writeStream = fs.createWriteStream(fullFilePath);
            yield new Promise((resolve, reject) => {
                writeStream
                    .once('close', resolve)
                    .once('error', reject);
                this.pipe(writeStream);
            });
        });
    }
    toBase64() {
        return __awaiter(this, void 0, void 0, function* () {
            const buffer = yield this.toBuffer();
            return buffer.toString('base64');
        });
    }
    /**
     * dataUrl: `data:image/png;base64,${base64Text}',
     */
    toDataURL() {
        return __awaiter(this, void 0, void 0, function* () {
            const base64Text = yield this.toBase64();
            if (!this.mimeType) {
                throw new Error('no mimeType');
            }
            const dataUrl = [
                'data:',
                this.mimeType,
                ';base64,',
                base64Text,
            ].join('');
            return dataUrl;
        });
    }
    toBuffer() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.boxType === file_box_type_1.FileBoxType.Buffer) {
                if (!this.buffer) {
                    throw new Error('no buffer!');
                }
                return this.buffer;
            }
            const stream = new stream_1.PassThrough();
            this.pipe(stream);
            const buffer = yield misc_1.streamToBuffer(stream);
            return buffer;
        });
    }
}
exports.FileBox = FileBox;
exports.default = FileBox;
